python/cudf/cudf/core/udf/utils.py

```diff
@@ -245,20 +245,22 @@ def _mask_get(mask, pos):
     return (mask[pos // MASK_BITSIZE] >> (pos % MASK_BITSIZE)) & 1
 
 
-def _generate_cache_key(frame, func: Callable, suffix="__APPLY_UDF"):
+def _generate_cache_key(frame, func: Callable, args, suffix="__APPLY_UDF"):
     """Create a cache key that uniquely identifies a compilation.
 
     A new compilation is needed any time any of the following things change:
     - The UDF itself as defined in python by the user
     - The types of the columns utilized by the UDF
     - The existence of the input columns masks
     """
+    scalar_argtys = tuple(typeof(arg) for arg in args)
```

Comment:
Are `args` always scalars? Are there supported argument types that are unknown to `typeof`? (basic numerical types and strings appear to succeed, but lists, dicts, custom classes `class A: pass`, and `pd.Timestamp` fail).

```suggestion
    scalar_types = tuple(typeof(arg) for arg in args)
```

Reply:
We only support scalar args. But you're right there should be a check for this way further up the pipeline. Happy to add it as part of this PR.

Reply:
addressed in https://github.com/rapidsai/cudf/pull/13194/commits/f85ee4d45daca2958a080d743e69ad9de1918b21